Native Half on ARM

mikekgfb opened the pull request:
Summary: Native Half on ARM

Test Plan: sandcastle

Differential Revision: D53585776

jgong5 commented on c10/util/Half-inl.h:
```diff
@@ -8,23 +8,41 @@
 #include <ATen/cpu/vec/vec_half.h>
 #endif
 
+// including this stanza seemed like good hygiene, but it ends up messing
+// with the compiler's ability to figure out what the execution mode is below
+#if 0
+#ifndef __NOT_CPU__
+#if defined(__GNUC__) || defined(__clang__)
+#if defined(__aarch64__)
+#if (!defined(__ARM_32BIT_STATE)) || (__ARM_32BIT_STATE != 1)
+#define NATIVE_FP16 1
+#endif // __ARM_32BIT_STATE
+#endif // __aarch64__
+#endif // GNUC or clang
+#endif // __NOT_CPU__
+#endif
```
What's the problem here?

mikekgfb replied:
If Half-inl is included directly (not such inclusion currently in the code base), then the symbols defined in Half.h are not available.

jgong5 commented on c10/util/Half.h:
```diff
@@ -289,6 +314,12 @@ C10_HOST_DEVICE inline float fp16_ieee_to_fp32_value(uint16_t h) {
+#if defined(NATIVE_FP16)
+  _Float16 fp16 = static_cast<_Float16>(f);
+  return reinterpret_cast<uint16_t>(fp16);
+#elif defined(X86_F16)
+  return _cvtss_sh(f, _MM_FROUND_TO_NEAREST_INT);
```
Where is `X86_F16` defined? I guess you need `at::vec::float2half_scalar(x)` here?

jgong5 commented on c10/util/Half.h:
```diff
@@ -322,12 +353,18 @@ inline uint16_t fp16_ieee_from_fp32_value(float f) {
-  unsigned short x;
+  union {
+#ifdef NATIVE_FP16
+    _Float16 y;
```
Does it need to specify `-mfp16-format` to enable `_Float16`? See https://gcc.gnu.org/onlinedocs/gcc/Half-Precision.html

ezyang commented:
Uhh, is sandcastle actually going to be a sufficient test for this

ezyang commented on c10/util/Half.h:
```diff
@@ -33,18 +33,32 @@
 
 #ifdef __CUDACC__
 #include <cuda_fp16.h>
+#define __NOT_CPU__
```
Since this is a header, might be better to use a macro that has TORCH or C10 in it to make it less likely to clobber someone else

ezyang commented on c10/util/Half.h:
```diff
@@ -33,18 +33,32 @@
+#if defined(__GNUC__) || defined(__clang__)
+#if defined(__aarch64__)
+#if (!defined(__ARM_32BIT_STATE)) || (__ARM_32BIT_STATE != 1)
+#define NATIVE_FP16 1
```
Ditto this macro, this macro is going to leak into PyTorch's public C API

ezyang reviewed (approved):
sure why not

Commit:
Native Half on ARM (#119483)

Summary:

Native Half on ARM

Test Plan: sandcastle

Differential Revision: D53585776

mikekgfb commented:
@pytorchbot merge

mikekgfb commented:
@pytorchbot merge

xwang233 commented:
cc @ptrblck @eqy, we may see an ARM build failure after this

malfet commented:
@pytorchbot revert -m "Broke nightly arm builds (and will be breaking runtime), as F16 arithmetic is ARMv8.2 only, see https://github.com/pytorch/pytorch/actions/runs/8000968963/job/21851281141" -c nosignal

mikekgfb commented:
@snadampal FYI
